src/PointwiseFunctions/GeneralRelativity/ComputeGhQuantities.cpp

```diff
@@ -167,52 +172,130 @@ tnsr::ijj<DataType, SpatialDim, Frame> deriv_spatial_metric(
+tnsr::ii<DataType, SpatialDim, Frame> time_deriv_of_spatial_metric(
+    const Scalar<DataType>& lapse,
+    const tnsr::I<DataType, SpatialDim, Frame>& shift,
+    const tnsr::iaa<DataType, SpatialDim, Frame>& phi,
+    const tnsr::aa<DataType, SpatialDim, Frame>& pi) noexcept {
+  tnsr::ii<DataType, SpatialDim, Frame> dt_spatial_metric {};
+  GeneralizedHarmonic::time_deriv_of_spatial_metric<SpatialDim, Frame,
+    DataType>(make_not_null(&dt_spatial_metric), lapse, shift, phi, pi);
   return dt_spatial_metric;
 }
 
+namespace {
 template <size_t SpatialDim, typename Frame, typename DataType>
-tnsr::a<DataType, SpatialDim, Frame> spacetime_deriv_of_det_spatial_metric(
+struct D4gBuffer;
+
+template <size_t SpatialDim, typename Frame>
+struct D4gBuffer<SpatialDim, Frame, double> {
+  // NOLINTNEXTLINE(cppcoreguidelines-pro-type-member-init)
+  explicit D4gBuffer(const size_t /*size*/) noexcept {}
+
+  tnsr::ijj<double, SpatialDim, Frame> deriv_of_g;
+  Scalar<double> det_spatial_metric;
+};
+
+template <size_t SpatialDim, typename Frame>
+struct D4gBuffer<SpatialDim, Frame, DataVector> {
+  private:
+    // We make one giant allocation so that we don't thrash the heap.
+    Variables<tmpl::list<::Tags::Tempijj<0, SpatialDim, Frame, DataVector>,
+      ::Tags::TempScalar<1, DataVector>>> buffer_;
+
+  public:
+    explicit D4gBuffer(const size_t size) noexcept : buffer_(size),
+      deriv_of_g(get<::Tags::Tempijj<0, SpatialDim, Frame, DataVector>>
+        (buffer_)),
+      det_spatial_metric(get<::Tags::TempScalar<1, DataVector>>(buffer_)) {}
+
+    tnsr::ijj<DataVector, SpatialDim, Frame>& deriv_of_g;
+    Scalar<DataVector>& det_spatial_metric;
+};
+}  // namespace
+
+template <size_t SpatialDim, typename Frame, typename DataType>
+void spacetime_deriv_of_det_spatial_metric(
+    const gsl::not_null<tnsr::a<DataType, SpatialDim, Frame>*>
+        d4_det_spatial_metric,
     const Scalar<DataType>& sqrt_det_spatial_metric,
     const tnsr::II<DataType, SpatialDim, Frame>& inverse_spatial_metric,
     const tnsr::ii<DataType, SpatialDim, Frame>& dt_spatial_metric,
     const tnsr::iaa<DataType, SpatialDim, Frame>& phi) noexcept {
-  auto d4_det_spatial_metric = make_with_value<tnsr::a<DataType,
-      SpatialDim, Frame>>(get(sqrt_det_spatial_metric), 0.);
-  const auto deriv_of_g = deriv_spatial_metric(phi);
-  const auto det_spatial_metric = square(get(sqrt_det_spatial_metric));
+  if (get_size(get<0>(*d4_det_spatial_metric)) !=
+      get_size(get(sqrt_det_spatial_metric))) {
+    tnsr::a<DataType, SpatialDim, Frame> d4_det_spatial_metric_(get(
+      sqrt_det_spatial_metric));
+    *d4_det_spatial_metric = d4_det_spatial_metric_;
+  }
+  auto& d4_g = *d4_det_spatial_metric;
+  // Use a Variables to reduce total number of allocations. This is especially
+  // important in a multithreaded environment.
+  D4gBuffer<SpatialDim, Frame, DataType>
+    buffer(get_size(get(sqrt_det_spatial_metric)));
```

Comment:
As in #1177, this does not save any allocations.  Perhaps add a return-by-reference form of `deriv_spatial_metric`?

Reply:
added a return-by-reference form of `deriv_spatial_metric`